File: ios/App/App/MusicFoldersPlugin.swift
import AVFoundation
import Capacitor
import Foundation
import UIKit
import UniformTypeIdentifiers

/// The music a person chose in the Files app, for Spotifie's Local Music.
///
/// The person picks a folder - or, where a Files provider will not list a
/// folder, some songs - in the system document picker. iOS grants access to
/// exactly that, and the plugin keeps a security-scoped bookmark so the grant
/// survives a restart. No permission prompt is involved and none is declared:
/// nothing outside what was chosen is reachable.
///
/// What this never does: write, move or delete a file, copy audio anywhere,
/// encode it, or send it anywhere. It lists files, reads their tags, and hands
/// back file addresses the web view plays in place. The bookmarks stay in the
/// app's own container, excluded from backup, and are never sent anywhere.
///
/// The same JavaScript interface as the Android plugin of the same name
/// (android/app/src/main/java/app/spotifie/android/MusicFoldersPlugin.java), so
/// one library (js/nativeLibrary.js) drives both.
///
/// NOT YET COMPILED OR RUN: written on Windows. Build and test it on a Mac
/// before relying on it - see P15_MAC_HANDOFF.private.md.
@objc(MusicFoldersPlugin)
public class MusicFoldersPlugin: CAPPlugin, CAPBridgedPlugin, UIDocumentPickerDelegate {
    public let identifier = "MusicFoldersPlugin"
    public let jsName = "MusicFolders"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "pickFolder", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "pickFiles", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listFolders", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "releaseFolder", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "scanFolder", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readTags", returnType: CAPPluginReturnPromise)
    ]

    /// What counts as music: the same list every other part of Spotifie uses.
    /// Which of these the web view can actually play is decided by the page.
    private static let audio: Set<String> = ["mp3", "flac", "wav", "m4a", "aac", "ogg", "opus"]

    /// A folder chosen by mistake (a whole drive) must not become an endless walk.
    private static let maxDepth = 8
    private static let maxFiles = 5000

    /// One thing the person chose: a folder, or a set of songs.
    private struct Source: Codable {
        let id: String
        var name: String
        let kind: String
        var bookmarks: [Data]
    }

    private var sources: [Source] = []
    /// The addresses whose security-scoped access is open, per source. Kept
    /// open while the app runs, so the web view can read the files it plays.
    private var opened: [String: [URL]] = [:]
    private var pendingCall: PluginCall?
    private var pendingKind = "folder"
    private let queue = DispatchQueue(label: "app.spotifie.ios.music-folders")

    override public func load() {
        queue.async {
            self.sources = self.readSources()
            for source in self.sources {
                self.openAccess(source)
            }
        }
    }

    // ============================================
    // Choosing, listing and forgetting
    // ============================================

    @objc func pickFolder(_ call: PluginCall) {
        present(call, kind: "folder")
    }

    @objc func pickFiles(_ call: PluginCall) {
        present(call, kind: "files")
    }

    private func present(_ call: PluginCall, kind: String) {
        DispatchQueue.main.async {
            guard self.pendingCall == nil else {
                call.reject("A picker is already open.", "BUSY")
                return
            }

            let types: [UTType] = kind == "folder" ? [.folder] : [.audio]
            let picker = UIDocumentPickerViewController(forOpeningContentTypes: types, asCopy: false)
            picker.allowsMultipleSelection = kind == "files"
            picker.delegate = self

            self.pendingCall = call
            self.pendingKind = kind
            self.bridge?.viewController?.present(picker, animated: true)
        }
    }

    public func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        pendingCall?.reject("Nothing was chosen.", "CANCELLED")
        pendingCall = nil
    }

    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let call = pendingCall else { return }
        pendingCall = nil
        let kind = pendingKind

        queue.async {
            var bookmarks: [Data] = []
            for url in urls {
                let started = url.startAccessingSecurityScopedResource()
                defer {
                    if started { url.stopAccessingSecurityScopedResource() }
                }
                if let data = try? url.bookmarkData(options: [], includingResourceValuesForKeys: nil, relativeTo: nil) {
                    bookmarks.append(data)
                }
            }

            guard !bookmarks.isEmpty else {
                call.reject("iOS did not grant lasting access to that choice.", "NOT_PERSISTED")
                return
            }

            // The same folder chosen again is the same folder, with its grant renewed.
            if kind == "folder", let chosen = urls.first,
               let index = self.sources.firstIndex(where: { $0.kind == "folder" && self.resolve($0).first?.standardizedFileURL.path == chosen.standardizedFileURL.path }) {
                self.closeAccess(self.sources[index].id)
                self.sources[index].bookmarks = bookmarks
                self.writeSources()
                self.openAccess(self.sources[index])
                call.resolve(["uri": self.sources[index].id, "name": self.sources[index].name])
                return
            }

            let name: String
            if kind == "folder" {
                name = urls.first?.lastPathComponent ?? "Music"
            } else {
                name = urls.count == 1 ? urls[0].deletingPathExtension().lastPathComponent : "\(urls.count) chosen songs"
            }

            let source = Source(id: (kind == "folder" ? "ios-folder:" : "ios-files:") + UUID().uuidString, name: name, kind: kind, bookmarks: bookmarks)
            self.sources.append(source)
            self.writeSources()
            self.openAccess(source)
            call.resolve(["uri": source.id, "name": source.name])
        }
    }

    /// What iOS still lets the app read.
    @objc func listFolders(_ call: PluginCall) {
        queue.async {
            let folders: [[String: Any]] = self.sources.compactMap { source in
                guard let urls = self.opened[source.id], !urls.isEmpty else { return nil }
                return ["uri": source.id, "name": source.name]
            }
            call.resolve(["folders": folders])
        }
    }

    /// Give a choice back. The files are not touched.
    @objc func releaseFolder(_ call: PluginCall) {
        guard let id = call.getString("uri") else {
            call.reject("Which folder?", "MISSING_URI")
            return
        }
        queue.async {
            self.closeAccess(id)
            self.sources.removeAll { $0.id == id }
            self.writeSources()
            call.resolve()
        }
    }

    // ============================================
    // Reading what was chosen
    // ============================================

    /// Every audio file in one choice: where it is inside it, its size and
    /// when it last changed - enough for the page to tell an unchanged song
    /// from a new one without opening it.
    @objc func scanFolder(_ call: PluginCall) {
        guard let id = call.getString("uri") else {
            call.reject("Which folder?", "MISSING_URI")
            return
        }
        queue.async {
            guard let source = self.sources.first(where: { $0.id == id }), let roots = self.opened[id], !roots.isEmpty else {
                call.reject("Spotifie no longer has access to that folder. Choose it again.", "NOT_GRANTED")
                return
            }

            var files: [[String: Any]] = []
            if source.kind == "folder", let root = roots.first {
                self.walk(root, into: &files)
            } else {
                for url in roots where files.count < MusicFoldersPlugin.maxFiles {
                    if let entry = self.describe(url, path: url.lastPathComponent) { files.append(entry) }
                }
            }
            call.resolve(["files": files])
        }
    }

    private func walk(_ root: URL, into files: inout [[String: Any]]) {
        let keys: [URLResourceKey] = [.isDirectoryKey, .fileSizeKey, .contentModificationDateKey]
        guard let enumerator = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: keys,
            options: [.skipsHiddenFiles, .skipsPackageDescendants]
        ) else { return }

        let base = root.standardizedFileURL.path
        for case let url as URL in enumerator {
            if files.count >= MusicFoldersPlugin.maxFiles { break }
            if enumerator.level > MusicFoldersPlugin.maxDepth {
                enumerator.skipDescendants()
                continue
            }
            let values = try? url.resourceValues(forKeys: [.isDirectoryKey])
            if values?.isDirectory == true { continue }

            var path = url.standardizedFileURL.path
            if path.hasPrefix(base + "/") { path = String(path.dropFirst(base.count + 1)) }
            if let entry = describe(url, path: path) { files.append(entry) }
        }
    }

    private func describe(_ url: URL, path: String) -> [String: Any]? {
        guard MusicFoldersPlugin.audio.contains(url.pathExtension.lowercased()) else { return nil }
        let values = try? url.resourceValues(forKeys: [.fileSizeKey, .contentModificationDateKey])
        return [
            "uri": url.absoluteString,
            "path": path,
            "name": url.lastPathComponent,
            "size": values?.fileSize ?? 0,
            "lastModified": Int64((values?.contentModificationDate?.timeIntervalSince1970 ?? 0) * 1000)
        ]
    }

    /// The tags of one song, read only for a song that is new or has changed.
    /// Only files inside something the person chose are read.
    @objc func readTags(_ call: PluginCall) {
        guard let value = call.getString("uri"), let url = URL(string: value), url.isFileURL else {
            call.reject("Which file?", "MISSING_URI")
            return
        }
        queue.async {
            guard self.insideGrant(url) else {
                call.reject("That file is not in something you chose.", "NOT_GRANTED")
                return
            }

            let asset = AVURLAsset(url: url)
            var out: [String: Any] = [:]
            for item in asset.commonMetadata {
                guard let key = item.commonKey, let text = item.stringValue?.trimmingCharacters(in: .whitespacesAndNewlines), !text.isEmpty else { continue }
                switch key {
                case .commonKeyTitle: out["title"] = text
                case .commonKeyArtist: out["artist"] = text
                case .commonKeyAlbumName: out["album"] = text
                default: break
                }
            }
            let seconds = CMTimeGetSeconds(asset.duration)
            if seconds.isFinite && seconds > 0 { out["duration"] = seconds }
            call.resolve(out)
        }
    }

    // ============================================
    // Bookmarks and access
    // ============================================

    private func resolve(_ source: Source) -> [URL] {
        return source.bookmarks.compactMap { data in
            var stale = false
            return try? URL(resolvingBookmarkData: data, options: [], relativeTo: nil, bookmarkDataIsStale: &stale)
        }
    }

    /// Open security-scoped access to one choice, renewing a stale bookmark.
    private func openAccess(_ source: Source) {
        var urls: [URL] = []
        var renewed = source.bookmarks
        var changed = false

        for (index, data) in source.bookmarks.enumerated() {
            var stale = false
            guard let url = try? URL(resolvingBookmarkData: data, options: [], relativeTo: nil, bookmarkDataIsStale: &stale) else { continue }
            guard url.startAccessingSecurityScopedResource() else { continue }
            urls.append(url)
            if stale, let fresh = try? url.bookmarkData(options: [], includingResourceValuesForKeys: nil, relativeTo: nil) {
                renewed[index] = fresh
                changed = true
            }
        }

        opened[source.id] = urls
        if changed, let index = sources.firstIndex(where: { $0.id == source.id }) {
            sources[index].bookmarks = renewed
            writeSources()
        }
    }

    private func closeAccess(_ id: String) {
        opened[id]?.forEach { $0.stopAccessingSecurityScopedResource() }
        opened[id] = nil
    }

    private func insideGrant(_ file: URL) -> Bool {
        let path = file.standardizedFileURL.path
        for (_, urls) in opened {
            for url in urls {
                let root = url.standardizedFileURL.path
                if path == root || path.hasPrefix(root + "/") { return true }
            }
        }
        return false
    }

    // ============================================
    // What is kept
    // ============================================

    /// Application Support/SpotifieMusicSources.json: bookmarks only, never
    /// audio and never a copy of a file. Excluded from iCloud and device backup.
    private func storeURL() -> URL? {
        guard let support = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else { return nil }
        try? FileManager.default.createDirectory(at: support, withIntermediateDirectories: true)
        return support.appendingPathComponent("SpotifieMusicSources.json")
    }

    private func readSources() -> [Source] {
        guard let url = storeURL(), let data = try? Data(contentsOf: url) else { return [] }
        return (try? JSONDecoder().decode([Source].self, from: data)) ?? []
    }

    private func writeSources() {
        guard var url = storeURL(), let data = try? JSONEncoder().encode(sources) else { return }
        try? data.write(to: url, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
        var values = URLResourceValues()
        values.isExcludedFromBackup = true
        try? url.setResourceValues(values)
    }
}

File: android/app/src/main/java/app/spotifie/android/MusicFoldersPlugin.java
package app.spotifie.android;

import android.app.Activity;
import android.content.ContentResolver;
import android.content.Intent;
import android.content.UriPermission;
import android.database.Cursor;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.provider.DocumentsContract;
import android.provider.DocumentsContract.Document;

import androidx.activity.result.ActivityResult;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

/**
 * The music folders a person chose, through Android's Storage Access Framework.
 *
 * The person picks a folder in the system's own picker; the app keeps a
 * persistable READ grant for that folder and nothing else. No storage
 * permission is declared or requested: nothing outside a chosen folder is
 * reachable, and a grant can be taken back from the system settings at any
 * time.
 *
 * What this never does: write, move or delete a file, copy audio anywhere,
 * encode it, or send it anywhere. It lists documents, reads their tags, and
 * hands back content:// addresses the WebView plays in place.
 *
 * Every method runs on Capacitor's plugin thread, never the UI thread.
 */
@CapacitorPlugin(name = "MusicFolders")
public class MusicFoldersPlugin extends Plugin {

    /** What counts as music: the same list every other part of Spotifie uses. */
    private static final Set<String> AUDIO = new HashSet<>(Arrays.asList("mp3", "flac", "wav", "m4a", "aac", "ogg", "opus"));

    /** A folder chosen by mistake (a whole drive) must not become an endless walk. */
    private static final int MAX_DEPTH = 8;
    private static final int MAX_FILES = 5000;

    // ============================================
    // Choosing, listing and forgetting folders
    // ============================================

    @PluginMethod
    public void pickFolder(PluginCall call) {
        Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        intent.addFlags(
            Intent.FLAG_GRANT_READ_URI_PERMISSION |
            Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION |
            Intent.FLAG_GRANT_PREFIX_URI_PERMISSION
        );
        startActivityForResult(call, intent, "folderPicked");
    }

    @ActivityCallback
    private void folderPicked(PluginCall call, ActivityResult result) {
        if (call == null) return;

        Intent data = result.getData();
        if (result.getResultCode() != Activity.RESULT_OK || data == null || data.getData() == null) {
            call.reject("No folder was chosen.", "CANCELLED");
            return;
        }

        Uri tree = data.getData();
        try {
            // Read only, and kept across restarts. Never write.
            resolver().takePersistableUriPermission(tree, Intent.FLAG_GRANT_READ_URI_PERMISSION);
        } catch (SecurityException e) {
            call.reject("Android did not grant lasting access to that folder.", "NOT_PERSISTED");
            return;
        }

        JSObject out = new JSObject();
        out.put("uri", tree.toString());
        out.put("name", displayName(tree));
        call.resolve(out);
    }

    /** The folders Android still lets the app read. */
    @PluginMethod
    public void listFolders(PluginCall call) {
        JSArray folders = new JSArray();
        for (UriPermission permission : resolver().getPersistedUriPermissions()) {
            Uri uri = permission.getUri();
            if (!permission.isReadPermission() || !DocumentsContract.isTreeUri(uri)) continue;

            JSObject folder = new JSObject();
            folder.put("uri", uri.toString());
            folder.put("name", displayName(uri));
            folders.put(folder);
        }

        JSObject out = new JSObject();
        out.put("folders", folders);
        call.resolve(out);
    }

    /** Give a folder back. The files in it are not touched. */
    @PluginMethod
    public void releaseFolder(PluginCall call) {
        String value = call.getString("uri");
        if (value == null) {
            call.reject("Which folder?", "MISSING_URI");
            return;
        }

        try {
            resolver().releasePersistableUriPermission(Uri.parse(value), Intent.FLAG_GRANT_READ_URI_PERMISSION);
        } catch (SecurityException ignored) {
            // Already gone, which is what was asked for.
        }
        call.resolve();
    }

    // ============================================
    // Reading a folder
    // ============================================

    /**
     * Every audio document inside one chosen folder: where it is inside the
     * folder, its size and when it last changed - enough for the page to tell
     * an unchanged song from a new one without opening it.
     */
    @PluginMethod
    public void scanFolder(PluginCall call) {
        Uri tree = grantedTree(call);
        if (tree == null) return;

        JSArray files = new JSArray();
        walk(tree, DocumentsContract.getTreeDocumentId(tree), "", 0, files);

        JSObject out = new JSObject();
        out.put("files", files);
        call.resolve(out);
    }

    private void walk(Uri tree, String documentId, String prefix, int depth, JSArray files) {
        if (depth > MAX_DEPTH || files.length() >= MAX_FILES) return;

        Uri children = DocumentsContract.buildChildDocumentsUriUsingTree(tree, documentId);
        String[] columns = {
            Document.COLUMN_DOCUMENT_ID,
            Document.COLUMN_DISPLAY_NAME,
            Document.COLUMN_MIME_TYPE,
            Document.COLUMN_SIZE,
            Document.COLUMN_LAST_MODIFIED
        };

        try (Cursor cursor = resolver().query(children, columns, null, null, null)) {
            if (cursor == null) return;

            while (cursor.moveToNext() && files.length() < MAX_FILES) {
                String id = cursor.getString(0);
                String name = cursor.getString(1);
                String mime = cursor.getString(2);
                if (id == null || name == null || name.startsWith(".")) continue;

                String path = prefix.isEmpty() ? name : prefix + "/" + name;

                if (Document.MIME_TYPE_DIR.equals(mime)) {
                    walk(tree, id, path, depth + 1, files);
                    continue;
                }
                if (!AUDIO.contains(extension(name))) continue;

                JSObject file = new JSObject();
                file.put("uri", DocumentsContract.buildDocumentUriUsingTree(tree, id).toString());
                file.put("path", path);
                file.put("name", name);
                file.put("size", cursor.isNull(3) ? 0 : cursor.getLong(3));
                file.put("lastModified", cursor.isNull(4) ? 0 : cursor.getLong(4));
                files.put(file);
            }
        } catch (Exception ignored) {
            // A folder that cannot be read - the grant taken back, a card
            // removed - is as far as this goes. What was found is still good.
        }
    }

    /**
     * The tags of one song, read only for a song that is new or has changed.
     * Only documents inside a chosen folder are read.
     */
    @PluginMethod
    public void readTags(PluginCall call) {
        String value = call.getString("uri");
        if (value == null) {
            call.reject("Which file?", "MISSING_URI");
            return;
        }

        Uri uri = Uri.parse(value);
        if (!insideGrant(uri)) {
            call.reject("That file is not in a chosen folder.", "NOT_GRANTED");
            return;
        }

        JSObject out = new JSObject();
        MediaMetadataRetriever retriever = new MediaMetadataRetriever();
        try {
            retriever.setDataSource(getContext(), uri);
            putIfPresent(out, "title", retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_TITLE));
            putIfPresent(out, "artist", retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ARTIST));
            putIfPresent(out, "album", retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUM));
            putIfPresent(out, "albumArtist", retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_ALBUMARTIST));

            String duration = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION);
            if (duration != null) out.put("duration", Long.parseLong(duration) / 1000.0);
        } catch (Exception ignored) {
            // A file with no readable tags is named after itself by the page.
        } finally {
            try {
                retriever.release();
            } catch (Exception ignored) {
                // nothing left to release
            }
        }
        call.resolve(out);
    }

    // ============================================
    // Helpers
    // ============================================

    private ContentResolver resolver() {
        return getContext().getContentResolver();
    }

    /** The tree named in the call, if Android still lets the app read it. */
    private Uri grantedTree(PluginCall call) {
        String value = call.getString("uri");
        if (value == null) {
            call.reject("Which folder?", "MISSING_URI");
            return null;
        }

        Uri tree = Uri.parse(value);
        for (UriPermission permission : resolver().getPersistedUriPermissions()) {
            if (permission.isReadPermission() && permission.getUri().equals(tree)) return tree;
        }

        call.reject("Spotifie no longer has access to that folder. Choose it again.", "NOT_GRANTED");
        return null;
    }

    /** Is this document inside one of the folders the person chose? */
    private boolean insideGrant(Uri document) {
        String address = document.toString();
        for (UriPermission permission : resolver().getPersistedUriPermissions()) {
            if (!permission.isReadPermission()) continue;
            String tree = permission.getUri().toString();
            if (address.startsWith(tree + "/document/")) return true;
        }
        return false;
    }

    private String displayName(Uri tree) {
        try {
            Uri root = DocumentsContract.buildDocumentUriUsingTree(tree, DocumentsContract.getTreeDocumentId(tree));
            try (Cursor cursor = resolver().query(root, new String[] { Document.COLUMN_DISPLAY_NAME }, null, null, null)) {
                if (cursor != null && cursor.moveToFirst() && cursor.getString(0) != null) return cursor.getString(0);
            }
        } catch (Exception ignored) {
            // fall through to the address itself
        }
        String last = tree.getLastPathSegment();
        return last == null ? "Music" : last.substring(last.lastIndexOf(':') + 1);
    }

    private static String extension(String name) {
        int dot = name.lastIndexOf('.');
        return dot == -1 ? "" : name.substring(dot + 1).toLowerCase(Locale.ROOT);
    }

    private static void putIfPresent(JSObject out, String key, String value) {
        if (value != null && !value.trim().isEmpty()) out.put(key, value.trim());
    }
}
